src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -99,6 +100,19 @@ internal static SafeSslHandle AllocateSslContext(SslProtocols protocols, SafeX50
                     context.Dispose();
                     throw CreateSslException(SR.net_allocate_ssl_context_failed);
                 }
+
+                if (certHandle != null && !certHandle.IsInvalid)
+                {
+                    using (X509Certificate2 cert = new X509Certificate2(certHandle.DangerousGetHandle()))
```

Comment:
SecureChannel holds onto SafeFreeCredentials which holds onto the certHandle for the life of the call, but you're right actually Steve, there's nothing stopping the user calling SslStream.Dispose() during any of all of this, which will bubble to certHandle.Dispose() and make it fall over. SetSslCertificate further up doesn't have any protection either. 

If thread-safety was a design intention here, it might be an idea to either addref like you say or implement locks a bit further up stream for a more global coverage instead of trying to protect each interop call individually. For instance a `lock(something)` around SecureChannel's `Close`, `AcquireServerCredentials` and `AcquireClientCredentials`. That way the user's `Dispose` will block until these interop calls have completed, and it would protect all platforms. Alternatively, a reasonable place to do a "Unix-only" AddRef/Release on both the CertHandle and KeyHandle, is in the `AllocateSslContext` ctor. Seems like the Windows "Pal" side of credential acquisitions are tucked away in some `SSPIInterface` and probably not susceptible to the same race condition, in which case perhaps a unix-only safety check is the better option. 

Happy to toss in whatever -- just let me know what you guys decide.